src/nbl/video/IDescriptorPool.cpp

```diff
@@ -4,60 +4,63 @@
 namespace nbl::video
 {
 
-core::smart_refctd_ptr<IGPUDescriptorSet> IDescriptorPool::createDescriptorSet(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout)
+bool IDescriptorPool::createDescriptorSets(uint32_t count, const IGPUDescriptorSetLayout* const* layouts, core::smart_refctd_ptr<IGPUDescriptorSet>* output)
 {
-    if (!isCompatibleDevicewise(layout.get()))
-        return nullptr;
+    core::vector<SDescriptorOffsets> descriptorOffsets(count);
 
-    return createDescriptorSet_impl(std::move(layout));
+    for (uint32_t i = 0u; i < count; ++i)
+    {
+        if (!isCompatibleDevicewise(layouts[i]))
+            return false;
+        
+        descriptorOffsets[i] = allocateDescriptorOffsets(layouts[i]);
+    }
+
+    return createDescriptorSets_impl(count, layouts, descriptorOffsets.data(), output);
 }
 
-bool IDescriptorPool::freeDescriptorSets(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets)
+bool IDescriptorPool::reset()
 {
-    const bool allowsFreeingDescriptorSets = m_flags & IDescriptorPool::ECF_FREE_DESCRIPTOR_SET_BIT;
-    if (!allowsFreeingDescriptorSets)
-        return false;
-
-    for (auto i = 0u; i < descriptorSetCount; ++i)
+    const bool allowsFreeing = m_creationParameters.flags.hasFlags(ECF_FREE_DESCRIPTOR_SET_BIT);
+    for (uint32_t t = 0; t < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++t)
     {
-        for (auto t = 0u; t < asset::EDT_COUNT; ++i)
-        {
-            const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(t);
+        if (!m_descriptorAllocators[t])
+            continue;
 
-            const uint32_t allocatedOffset = descriptorSets[i]->getDescriptorStorageOffset(type);
-            if (allocatedOffset == ~0u)
-                continue;
+        const uint32_t allocatedCount = m_descriptorAllocators[t]->getAllocatedDescriptorCount(allowsFreeing);
+        std::destroy_n(getDescriptorStorage(static_cast<asset::IDescriptor::E_TYPE>(t)), allocatedCount);
```

Comment:
For a PoolAddressAllocator (`allowsFreeing==true`) the range of _constructed_ items will not be contiguous (if you free a non-last descriptor set), and this will blow up

Reply:
You can only do this trick `if (allowsFreeing)` and therefore you know the allocator to be linear.

Otherwise you'd have to do through each allocated descriptor set and make it free the range assigned to it.

This would be possible if you kept a `core::unique_ptr<IGPUDescriptorSet*[]> m_allocatedDescriptors;` and a PoolAddressAllocator to give each newly allocated descriptor set an ID.

Reply:
you'd obviously `fill_n(m_allocatedDescriptors,m_creationParameters.maxSets,nullptr)` and reset the pool address allocator at the end of a `reset()`

Reply:
Not that the `m_allocatedDescriptors` array is an array of _dumb_ pointers to avoid circular references.

Furthermore thanks to this you'd be able to manually mark `IGPUDescriptorSets` as zombies and you wouldn't need `m_version` anymore

Reply:
Basically
```cpp
if (m_creationParameters.flags.hasFlags(ECF_FREE_DESCRIPTOR_SET_BIT))
{
    for (uint32_t t = 0; t < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++t)
    {
        if (!m_descriptorAllocators[t])
            continue;
        const uint32_t allocatedCount = m_descriptorAllocators[t]->getAllocatedDescriptorCount();
        std::destroy_n(getDescriptorStorage(static_cast<asset::IDescriptor::E_TYPE>(t)), allocatedCount);
    }
}
else
{
    for (uint32_t i=0u; i<m_CreationParameters.maxSets; i++)
    if (m_allocatedDescriptorSets[i])
        deleteSetStorage(m_allocatedDescriptorSets[i]); // basically what `~IGPUDescriptorSet` calls minus the free of the set itself
}
return reset_impl();
```

Reply:
There's even a [IteratablePoolAddressAllocator.h](https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/b3719ddb1377d8af8f8d5f4dd112d3c844394553..9128500897be7e847f13ed37e55958948225d69d#diff-4a46350337b8d944636270cbad79d112989df3e0778350c4dd47df5315c79bd8) that would let you iterate directly over the non-null `m_allocatedDescriptorSets`

Reply:
Actually if we support "immediate smarpointer destruction" due to our improved reading of the Vulkan spec, we're this is the correct approach always:
```cpp
for (const uint32_t setIndex : m_setAllocator)
{
    // basically what `~IGPUDescriptorSet` calls minus the free-ing of the set itself
    deleteSetStorage(m_allocatedDescriptorSets[setIndex]);
    // dont make zombies keep the parent alive
    m_allocatedDescriptorSets[setIndex]->m_parentPool = nullptr;
}
m_setAllocator.reset();
return reset_impl();
```

because

Reply:
`IteratablePoolAddressAllocator` doesn't seem to have a `public` and non-`const` `begin` method, I'll make one.

Reply:
Resolved.